File: leetcode/p00061/p00061.go
package p00061

type ListNode struct {
	Val  int
	Next *ListNode
}

// Time Complexity: O(n)
// Space Complexity: O(n)
func rotateRight(head *ListNode, k int) *ListNode {
	if head == nil {
		return nil
	}

	size := 0
	current := head
	for current != nil {
		size++
		current = current.Next
	}

	k %= size

	tail := head
	for i := 0; i < k; i++ {
		tail = tail.Next
	}

	anchor := head
	for tail.Next != nil {
		tail = tail.Next
		anchor = anchor.Next
	}

	tail.Next = head
	new_head := anchor.Next
	anchor.Next = nil

	return new_head
}

File: leetcode/p00061/p00061.py
from typing import Optional, List

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    # Time Complexity: O(n)
    # Space Complexity: O(n)
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head:
            return None
        
        size = 0
        current = head
        while current:
            size += 1
            current = current.next
        
        k %= size

        tail = head
        for _ in range(k):
            tail = tail.next
        
        tmp = head
        while tail.next:
            tail = tail.next
            tmp = tmp.next
        
        tail.next = head
        new_head = tmp.next
        tmp.next = None

        return new_head

def getVals(head: Optional[ListNode]) -> List[int]:
    res = []

    while head:
        res.append(head.val)
        head = head.next

    return res


if __name__ == '__main__':
    s = Solution()

    list = [ListNode(1), ListNode(2), ListNode(3), ListNode(4), ListNode(5)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    assert getVals(s.rotateRight(list[0], 2)) == [4,5,1,2,3]

    list = [ListNode(0), ListNode(1), ListNode(2)]
    for i in range(0, len(list)-1):
        list[i].next = list[i+1]
    assert getVals(s.rotateRight(list[0], 4)) == [2,0,1]
